File: spring-data-requery/src/test/java/org/springframework/data/requery/domain/hierarchy/AbstractNodeAttribute.java
package org.springframework.data.requery.domain.hierarchy;

import io.requery.*;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.requery.domain.AbstractPersistable;
import org.springframework.data.requery.domain.ToStringBuilder;

import java.util.Objects;

@Getter
@Setter
@Entity
public class AbstractNodeAttribute extends AbstractPersistable<Long> {

    private static final long serialVersionUID = 7424000305448372730L;

    @Key
    @Generated
    @Column(name = "attr_id")
    protected Long id;

    @Column(name = "attr_name")
    protected String name;

    @Column(name = "attr_value")
    protected String value;

    @ManyToOne
    @ForeignKey
    protected AbstractTreeNode node;

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Transient
    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("name", name)
            .add("value", value);
    }
}

File: src/test/kotlin/com/coupang/springframework/data/requery/domain/hierarchy/AbstractNodeAttribute.kt
package com.coupang.springframework.data.requery.domain.hierarchy

import com.coupang.kotlinx.objectx.ToStringBuilder
import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*

/**
 * AbstractNodeAttribute
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
@Entity
abstract class AbstractNodeAttribute: AbstractPersistable<Long>() {

    @get:Key
    @get:Generated
    abstract override val id: Long?


    abstract var name: String?

    abstract var value: String?


    @get:ManyToOne(cascade = [CascadeAction.SAVE, CascadeAction.DELETE])
    @get:ForeignKey(referencedColumn = "nodeId")
    abstract var node: AbstractTreeNode?

    override fun hashCode(): Int {
        return name?.hashCode() ?: System.identityHashCode(this)
    }

    @Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("name", name)
            .add("value", value)
    }
}